torchrl/data/replay_buffers/rb_prototype.py

```diff
@@ -0,0 +1,177 @@
+import collections
+from concurrent.futures import ThreadPoolExecutor
+from typing import Any, Callable, Optional, Sequence, Union, Tuple
+
+import torch
+
+from ..tensordict.tensordict import TensorDictBase, LazyStackedTensorDict
+from .replay_buffers import pin_memory_output, stack_tensors, stack_td
+from .samplers import Sampler, RandomSampler
+from .storages import Storage, ListStorage
+from .utils import INT_CLASSES, to_numpy
+from .writers import Writer, RoundRobinWriter
+
+
+class ReplayBuffer:
+    def __init__(
+        self,
+        storage: Storage = None,
+        sampler: Sampler = None,
+        writer: Writer = None,
+        collate_fn: Callable = None,
+        pin_memory: bool = False,
+        prefetch: int = None,
+    ) -> None:
+        self._storage = storage or ListStorage(size=1_000)
+        self._sampler = sampler or RandomSampler()
+        self._writer = writer or RoundRobinWriter()
+        self._writer.register_storage(self._storage)
+
+        self._collate_fn = collate_fn or stack_tensors
+        self._pin_memory = pin_memory
+
+        self._prefetch = bool(prefetch)
+        self._prefetch_cap = prefetch or 0
+        self._prefetch_queue = collections.deque()
+        if self._prefetch_cap:
+            self._prefetch_executor = ThreadPoolExecutor(max_workers=self._prefetch_cap)
+
+    def __len__(self) -> int:
+        return len(self._storage)
```

Comment:
Ah yes - I removed the locks when I was writing the prototype. I just added the locks back in only now.

But I'd question the need for the lock here anyways?

If we go through a scenario in which it's important for the `len(rb)` to return **exactly** the right length, if we do something like:

```python
if len(rb) == x:
    # Already here - there's no guarantee that len(rb) is the same as on the line above
    # regardless of whether we stick a lock in `ReplayBuffer.__len__` or not, right?
    do_something()
```

and so if you want to use `len()` for something important - you'll need to lock it from the outside anyways, or if you just use it for logging - an approximation (without lock) is probably fine?

In cases where we actually do want to make sure we check length of the RB, and then act on that atomically - do we want to expose the internal `ReplayBuffer._replay_lock` to the RB users? So they can do stuff like:

```python
with rb.replay_lock:
    if len(rb) == 10:
        do_something_that_is_possible_only_when_rb_is_of_size_10_exactly()
```

Let me know if you think I'm missing something here - I'll be happy to add the lock there if it's actually needed.

Reply:
Great points. I wanted to have a lock from the pov of correctness (since we have multiple writers here). Let me think more about if a lock is really needed. For the second part (i.e.  expose the internal ReplayBuffer._replay_lock), that is just a regular lock correct? Is there any difference between using this lock vs creating our own, high-level lock ?

Reply:
I think the only difference would be performance - using an internal `ReplayBuffer._replay_lock` means that it is used only for internal `ReplayBuffer`'s operations that need to be under lock, while if we use our own lock instead - we are forced to lock for the entire duration for any `ReplayBuffer` method. One example of that would be the `ReplayBuffer.__getitem__()` which in very very simple terms does two things:

```python
class ReplayBuffer:
    def __getitem__(self, index):
        with self.replay_lock:
            access_storage()  # 1. accesses Storage
        collate_data()  # 2. collates data if necessary
```

now if from the outside you want to do something like:

```python
with rb.replay_lock():
    do_something()
```

the `do_something` will run as soon as the storage has been accessed in `__getitem__`, but it won't have to wait for the `collate_data()` to execute.

While if you lock everything from the outside with your own lock - you're forced to basically run everything sequentially:

```python
# thread 1
with lock():
    rb[10]
    
# thread 2
with lock():
    do_something()
```

In this case `lock` in thread 2 will only stop blocking after the entire `rb.__getitem__()` has executed, even though technically we don't need to wait for the `collate_data()` to `do_something()`.

This is a kind of dummy example, but hopefully it conveys what I mean. Let me know what do you think.